### ISSUE TITLE
SSHListener is not a context manager anymore (API break)

### ISSUE THREAD

**Issue by pocek** [ronf/asyncssh#236]
Hi. Before version 2.0.0 `SSHListener` was inheriting `asyncio.AbstractServer` and thanks to that it was possible to use it as a context manager (it was quite convenient). That's no longer the case since 9dc81c869d0a8108e39b8c42b2a8e0260c1e8187. This breaks the API and doesn't seem to be intentional.

**Comment by ronf** [comment]
You're right - this was not intentional. My apologies! I had no idea that AbstractServer implemented async context manager support, and am honestly a bit surprised that this is the case considering that all of the other methods on it are completely abstract (as the class name suggests).

I'm happy to add back this capability. I removed the inheritance only because I wanted to silence pylint warnings about missing methods added in Python 3.7 (get_loop, start_serving, is_serving, and serve_forever) that I hadn't implemented (and wasn't sure if I'd ever implement).

Actually, if I make the functions that return SSHListener objects into explicit async context managers using the decorator I use elsewhere, I can probably make the syntax a bit nicer. I imagine right now you'd need to write something like:

```python
    listener = await asyncssh.listen(...args...)
    async with listener:
        ...more code goes here...
```

Is that the kind of thing you're doing?

With my decorator, I could continue to support the above but also support the following:

```python
    async with asyncssh.listen(...args...) as listener:
        ...more code goes here...
```

**Comment by ronf** [comment]
After looking at this a bit more closely, I noticed that some of the functions returning SSHListener right now can return `None` on failure, rather than raising an exception, which makes it a bit uglier to try and use the return value as a context manager. I wish I had noticed this before releasing 2.0.0, as this was one of the API changes I made in 2.0.0 for the connect_agent() call. To do the same thing here, I'd need to make a similar change to SSHClientConnection.connect_server() and SSHClientConnection.create_unix_server() and the functions that call them (start_server, start_unix_server, forward_local_port, forward_local_path, forward_remote_port, forward_remote_path) to raise exceptions rather than returning None, and I sort of missed my change unless I want to make additional incompatible changes in the API.

Also, there's really no useful information I can put in the exception being raised in this case, since the remote SSH server just returns a failure response with no details about why it failed.

For now, I think I'll go ahead and restore the AsyncSSH 1.x behavior, but not actually make these functions support "async with" directly. However, suggestions are welcome.

Out of curiosity, what kind of code do you put in the "async with" block in this case, and how do you typically trigger exiting the block?

**Comment by ronf** [comment]
A fix for this is now available in the "develop" branch.

**Comment by ronf** [comment]
After sleeping on this, I've decided I'm going to look into making the changes to replace special return values used to signal errors with proper exceptions in AsyncSSH 2.0. Any cleanup of this sort will be limited to only AsyncSSH 2.0.x releases, and the API will stabilize again beginning in AsyncSSH 2.1. I'll try to keep any changes needed in calling code as simple as possible. I think the benefit of doing this now rather than waiting until AsyncSSH 3.x is worth relaxing the usual version numbering rules, especially considering AsyncSSH 2.0.0 came out only a few days ago and shouldn't have seen wide adoption yet.

**Comment by ronf** [comment]
This change is now available in the "develop" branch. The following methods were changes to raise a new ChannelListenError exception instead of returning None:

* SSHClientConnection.create_server
* SSHClientConnection.create_unix_server
* SSHClientConnection.start_server
* SSHClientConnection.start_unix_server
* SSHClientConnection.forward_remote_port
* SSHClientConnection.forward_remote_path

In addition, all of the following methods are now decorated as being able to return async context managers, so they can be directly called from within "async with" statements:

* create_server
* listen
* listen_reverse
* SSHClientConnection.create_server
* SSHClientConnection.create_unix_server
* SSHClientConnection.listen_ssh
* SSHClientConnection.listen_reverse_ssh
* SSHClientConnection.start_server
* SSHClientConnection.start_unix_server
* SSHClientConnection.forward_local_port
* SSHClientConnection.forward_local_path
* SSHClientConnection.forward_remote_port
* SSHClientConnection.forward_remote_path
* SSHClientConnection.forward_socks

**Comment by ronf** [comment]
These changes are now available in AsyncSSH 2.0.1.
